src/syntaxtree.c

```diff
@@ -0,0 +1,907 @@
+{
+    Obj  sel;
+    Obj  lev;
+    Obj  lst;
+    Obj  tmp;
+    UInt narg, i;
+
+    sel = SyntaxTreeCompiler(ARGI_INFO(stat, 1));
+    lev = SyntaxTreeCompiler(ARGI_INFO(stat, 2));
+
+    AssPRec(result, RNamName("sel"), sel);
+    AssPRec(result, RNamName("lev"), lev);
+
+    narg = NARG_SIZE_INFO(SIZE_STAT(stat)) - 2;
+    lst = NEW_PLIST(T_PLIST, narg);
+    SET_LEN_PLIST(lst, narg);
+
+    for (i = 1; i <= narg; i++) {
+        tmp = SyntaxTreeCompiler(ARGI_INFO(stat, i + 2));
+        SET_ELM_PLIST(lst, i, tmp);
+        CHANGED_BAG(lst);
+    }
+    AssPRec(result, RNamName("args"), lst);
+
+    return result;
+}
+
+static Obj SyntaxTreeElmList(Obj result, Stat stat)
+{
+    Obj list, refs;
+    Int nr;
+    Int i;
+
+    list = SyntaxTreeCompiler(ADDR_STAT(stat)[0]);
+    AssPRec(result, RNamName("list"), list);
+
+    nr = SIZE_STAT(stat) / sizeof(Stat) - 1;
+
+    refs = NEW_PLIST(T_PLIST, nr);
+    SET_LEN_PLIST(refs, nr);
+    AssPRec(result, RNamName("refs"), refs);
+
+    for(i = 1; i < 1 + nr; i++) {
+        SET_ELM_PLIST(refs, i, SyntaxTreeCompiler(ADDR_STAT(stat)[i]));
+        CHANGED_BAG(refs);
+    }
+    return result;
+}
+
+static Obj SyntaxTreeFunc(Obj result, Obj func)
+{
+    Obj str;
+    Obj stats;
+    Obj argnams;
+    Obj locnams;
+
+    Bag oldFrame;
+    Int narg;
+    Int nloc;
+    Int i;
+
+    if (NAME_FUNC(func)) {
+        AssPRec(result, RNamName("name"), NAME_FUNC(func));
+    }
+
+    narg = NARG_FUNC(func);
+    if (narg < 0) {
+        AssPRec(result, RNamName("variadic"), True);
+        narg = -narg;
+    }
+    else {
+        AssPRec(result, RNamName("variadic"), False);
+    }
+    AssPRec(result, RNamName("narg"), INTOBJ_INT(narg));
+
+    /* names of arguments */
+    argnams = NEW_PLIST(T_PLIST, narg);
+    SET_LEN_PLIST(argnams, narg);
+    AssPRec(result, RNamName("argnams"), argnams);
+    for (i = 1; i <= narg; i++) {
+        if (NAMI_FUNC(func, i) != 0) {
+            str = CopyToStringRep(NAMI_FUNC(func, i));
+        }
+        else {
+            /* TODO: Probably put the number in */
+            C_NEW_STRING_DYN(str, "localvar");
+        }
+        SET_ELM_PLIST(argnams, i, str);
+        CHANGED_BAG(argnams);
+    }
+
+    /* names of local variables */
+    nloc = NLOC_FUNC(func);
+    AssPRec(result, RNamName("nloc"), INTOBJ_INT(nloc));
+    locnams = NEW_PLIST(T_PLIST, nloc);
+    SET_LEN_PLIST(locnams, nloc);
+    AssPRec(result, RNamName("locnams"), locnams);
+    for (i = 1; i <= nloc; i++) {
+        if (NAMI_FUNC(func, narg + i) != 0) {
+            str = CopyToStringRep(NAMI_FUNC(func, narg + i));
+        }
+        else {
+            /* TODO: Probably put the number in */
+            C_NEW_STRING_DYN(str, "localvar");
+        }
+        SET_ELM_PLIST(locnams, i, str);
+        CHANGED_BAG(locnams);
+    }
+
+    /* switch to this function (so that 'ADDR_STAT' and 'ADDR_EXPR' work) */
+    SWITCH_TO_NEW_LVARS(func, narg, nloc, oldFrame);
```

Comment:
My suggestion (and one on which I'd be happy to work on, if people agree with the general idea) would be to *not* use `ADDR_STAT` nor `ADDR_EXPR` here. Instead, pass the `T_BODY` object as an additional argument `body` to each syntax tree callback; then replace `ADDR_STAT(stat)[i]` by `ADDR_OBJ(body)[stat+i]` -- and of course one can use helper functions or macros to make this nicer, and e.g. replace `ADDR_STAT(stat)[i]` by, say `ReadStat(body, stat, i)` with
```C
static inline Stat ReadStat(Obj body, Stat stat, UInt i)
{
   return ADDR_OBJ(body)[stat+i];
}
```

Some more work is needed (e.g. for `ARGI_INFO`), so it's not a quick change, but if one introduces the above `ReadStat` helper, it could easily be done incrementally. 

Added bonus: an error during syntax tree compilation does not risk leaving `STATE(PtrBody)` in a bad state.